.github/workflows/ci-base.yaml

```diff
@@ -13,31 +13,41 @@ defaults:
 jobs:
   lint-and-style:
     name: Static Analysis
-    runs-on: ubuntu-latest
+    runs-on: ubuntu-22.04
     steps:
       - name: Install dependencies
         run: |
           sudo DEBIAN_FRONTEND=noninteractive apt-get -qy update
           sudo DEBIAN_FRONTEND=noninteractive apt-get -qy install tox
       - name: Git checkout
         uses: actions/checkout@v2
-      - name: Linting and style
-        run: tox -e flake8 -e black -e isort
-      - name: mypy
+      - name: Formatting
+        run: tox -e black -e isort
+      - name: Mypy
         run: tox -e mypy
   unit-tests:
-    name: Unit
-    runs-on: ubuntu-18.04
+    name: Matrix
+    strategy:
+      matrix:
+        testenv:
+          - {os: ubuntu-18.04, pyver: py35, deadsnake: python3.5}
+          - {os: ubuntu-18.04, pyver: py36}
+          - {os: ubuntu-20.04, pyver: py38}
+          - {os: ubuntu-22.04, pyver: py310}
+    runs-on: ${{ matrix.testenv.os }}
     steps:
       - name: Install dependencies
         run: |
           sudo DEBIAN_FRONTEND=noninteractive apt-get -qy update
-          sudo DEBIAN_FRONTEND=noninteractive apt-get -qy install python3-venv
+          sudo DEBIAN_FRONTEND=noninteractive apt-get -qy install tox
+      - name: Install older Python from deadsnakes PPA
+        if: matrix.testenv.deadsnake != ''
+        run: |
+          sudo add-apt-repository --yes ppa:deadsnakes/ppa
+          sudo DEBIAN_FRONTEND=noninteractive apt-get -qy install "${{ matrix.testenv.deadsnake }}"
       - name: Git checkout
         uses: actions/checkout@v2
-      - name: Linting and style
-        run: |
-          python3 -m venv unit-test-venv
-          . unit-test-venv/bin/activate
-          pip install tox tox-pip-version
-          tox -e py3 -e py3-bionic -e py3-xenial
+      - name: Flake8
+        run: tox -e "${{ matrix.testenv.pyver }}-flake8"
+      - name: Unit
+        run: tox -e "${{ matrix.testenv.pyver }}-test"
```

Comment:
A single tox call (`tox -e pyX-flake8 -e pyX-test`) is slightly faster, but having envs in separate steps adds clarity to the results IMHO, so I think it's worth keeping it like this, especially given that the bulk of CI time is taken by the integration tests.

![image](https://user-images.githubusercontent.com/7614669/169270941-4ff36e15-53d5-4127-963e-a67349437fec.png)